cmd/pod-scaler/frontend.go

```diff
@@ -0,0 +1,431 @@
+package main
+
+import (
+	"embed"
+	_ "embed"
+	"encoding/json"
+	"fmt"
+	"io/fs"
+	"net/http"
+	"sort"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/openhistogram/circonusllhist"
+	"github.com/sirupsen/logrus"
+
+	corev1 "k8s.io/api/core/v1"
+	prowConfig "k8s.io/test-infra/prow/config"
+	"k8s.io/test-infra/prow/flagutil"
+	"k8s.io/test-infra/prow/interrupts"
+	"k8s.io/test-infra/prow/metrics"
+	"k8s.io/test-infra/prow/pjutil"
+	"k8s.io/test-infra/prow/simplifypath"
+
+	"github.com/openshift/ci-tools/pkg/api"
+	pod_scaler "github.com/openshift/ci-tools/pkg/pod-scaler"
+)
+
+// l keeps the tree legible
+func l(fragment string, children ...simplifypath.Node) simplifypath.Node {
+	return simplifypath.L(fragment, children...)
+}
+
+const (
+	OrgQuery       = "org"
+	RepoQuery      = "repo"
+	BranchQuery    = "branch"
+	VariantQuery   = "variant"
+	TargetQuery    = "target"
+	StepQuery      = "step"
+	BuildQuery     = "build"
+	ContainerQuery = "container"
+)
+
+// metadataQueryMapping defines how we expose a metadata entry to the user -
+// in many cases we can know that a specific metadata item only needs some
+// subset of fields to fully qualify it, so we will not expose all of them
+// to the user on the UI. Furthermore, we need to map those fields to query
+// parameters and expose that information to the UI as well.
+type metadataQueryMapping struct {
+	// matches determines if the given metadata matches this mapping
+	matches func(*pod_scaler.FullMetadata) bool
+
+	// preProcess acts on the metadata before the mapping is sent to the UI
+	preProcess func(*pod_scaler.FullMetadata)
+	// fields define how metadata is mapped to a query
+	fields []*fieldMapping
+	// postProcess acts on the metadata after the mapping is read from a request
+	postProcess func(*pod_scaler.FullMetadata)
+}
+
+type fieldMapping struct {
+	// query is the request query that should hold this field
+	query string
+	// field is the value the query should be inserted field
```

Comment:
This comment doesn't make sense to me

Reply:
there's some copy pasta nonsense there huh